concepts.tex

```diff
@@ -49,8 +47,27 @@
 function that returns a value may copy or move the returned object, or may return a
 reference. \newtxt{\enternote Regular functions may have side effects. \exitnote}
 \oldtxt{Regular functions are allowed to have side effects and preconditions that,
-when violated, make the function non-regular. \enternote A function that returns void
-is necessarily a regular function. \exitnote}
+when violated, make the function non-regular. \enternote A function that returns
+\tcode{void} is necessarily a regular function. \exitnote}
+
+\pnum
+\newtxt{Expressions appearing in a \techterm{requires-expression} in this document
+are required to be equality preserving, except for those annotated with the comment
+``not equality preserving.'' An expression so annotated may be equality preserving,
+but it is not required to be so.}
+
+\pnum
+\newtxt{Expressions required by this specification to be equality preserving are
+further required to be stable: two evaluations of such an expression with the same
+input objects must have the same results absent any intervening direct modification
+of those input objects. \enternote This requirement allows generic code to reason
+about the current values of objects based on knowledge of the prior values as
+observed via equality preserving expressions. It effectively forbids spontaneous
+changes to an object, changes to an object from another thread of execution, changes
+to an object as side effects of non-modifying actions, and changes to an object as
+side effects of modifying a distinct object if those changes could be observable
+to a library function via an equality preserving expression that is required to be
+valid for that object. \exitnote}
```

Comment:
Notes are non-normative. We don't have any normative description of "direct modification". I think we need one.